# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/SharedFontList.cpp
# Commit: 06c26e973285
# Full Hash: 06c26e97328589418681f58d637df751451f04ca
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2020-05-26 21:37:52
# Regressor Bug: 1640119
# File Overlap Count: 1
# Description:
#   Bug 1640119 - Pass shared-memory blocks for the font list as part of SetXPCOMProcessAttributes. r=jwatt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D76507
# ==============================================================================

diff -r fbdbe0cdce92 -r 06c26e973285 gfx/thebes/SharedFontList.cpp
--- a/gfx/thebes/SharedFontList.cpp	Tue May 26 11:38:58 2020 +0000
+++ b/gfx/thebes/SharedFontList.cpp	Tue May 26 09:30:17 2020 +0000
@@ -499,6 +499,25 @@
       MOZ_CRASH("parent: failed to initialize FontList");
     }
   } else {
+    // Initialize using the list of shmem blocks passed by the parent via
+    // SetXPCOMProcessAttributes.
+    auto& blocks = dom::ContentChild::GetSingleton()->SharedFontListBlocks();
+    for (auto handle : blocks) {
+      auto newShm = MakeUnique<base::SharedMemory>();
+      if (!newShm->IsHandleValid(handle)) {
+        // Bail out and let UpdateShmBlocks try to do its thing below.
+        break;
+      }
+      if (!newShm->SetHandle(handle, true)) {
+        MOZ_CRASH("failed to set shm handle");
+      }
+      if (!newShm->Map(SHM_BLOCK_SIZE) || !newShm->memory()) {
+        MOZ_CRASH("failed to map shared memory");
+      }
+      mBlocks.AppendElement(new ShmBlock(std::move(newShm)));
+    }
+    blocks.Clear();
+    // Update in case of any changes since the initial message was sent.
     for (unsigned retryCount = 0; retryCount < 3; ++retryCount) {
       if (UpdateShmBlocks()) {
         return;
@@ -589,6 +608,18 @@
   return true;
 }
 
+void FontList::ShareBlocksToProcess(nsTArray<base::SharedMemoryHandle>* aBlocks,
+                                    base::ProcessId aPid) {
+  MOZ_RELEASE_ASSERT(mReadOnlyShmems.Length() == mBlocks.Length());
+  for (auto& shmem : mReadOnlyShmems) {
+    base::SharedMemoryHandle* handle =
+        aBlocks->AppendElement(base::SharedMemory::NULLHandle());
+    if (!shmem->ShareToProcess(aPid, handle)) {
+      MOZ_CRASH("failed to share block");
+    }
+  }
+}
+
 // The block size MUST be sufficient to allocate the largest possible
 // SharedBitSet in a single contiguous block, following its own
 // Allocated() field.